### PR TITLE
Add email matcher

### PR DESCRIPTION
**Pull request by srushti**
```js
// match any valid email
id: 'email'
id: s.email()

// only match a specific domain
id: s.email({domain: 'example.org'})
```

@TabDigital/ping-api

### CODE REVIEW

File: test/matchers/email.spec.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+var email  = require('../../lib/matchers/email');
+
+describe('email matcher', function() {
+
+  it('matches email', function() {
+    email().match('path', 'foo@bar.com').should.eql([])
```
Comment (nguyenchr) [inline test/matchers/email.spec.js]: should we use the customer `should` assertions that the other specs are using?

**Comment by nguyenchr** [comment]
Coverage check is failing

File: test/matchers/email.spec.js (hunk 1 above)
Comment (srushti) [inline test/matchers/email.spec.js]: Fixed.

**Comment by nguyenchr** [comment]
:+1:

File: lib/matchers/email.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  },
+
+  match: function(path, value) {
+    if (typeof value !== 'string') return TYPE_ERROR;
+
+    if (value.length > 254)
+      return TYPE_ERROR;
+
+    if (!EMAIL_REGEX.test(value))
+      return TYPE_ERROR;
```
Comment (mehdivk) [inline lib/matchers/email.js]: Could be better to return in single line

File: lib/matchers/email.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    if (this.hasDomain && !this.domainRegex.test(value)) {
+      return 'should be a valid email address at ' + this.opts.domain;
+    }
+
+    // Further checking of some things regex can't handle
+    var parts = value.split("@");
+    if (parts[0].length > 64)
```
Comment (mehdivk) [inline lib/matchers/email.js]: Good to have `'` instead of `"`, that's the only place you used " in this pr

File: test/matchers/email.spec.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    invalidEmailId = randomString(64) + '@' + randomString(63) + '.' + randomString(63) + '.' + randomString(58) + '.com';
+    email().match('', invalidEmailId).should.have.error('should be a valid email address');
+  })
+
+  it('generates a string schema with pattern', function() {
+    var d = new email().toJSONSchema();
+    var reg = new RegExp(d.pattern, 'i');
+    d.type.should.equal('string');
+    reg.test('foo@bar.com').should.be.true;
+    reg.test('FoO.bAr1@baz.com').should.be.true;
```
Comment (mehdivk) [inline test/matchers/email.spec.js]: I remember @vickvu shared a note about problem with this type of assertion against different versions of shouldjs, upgrading to Should `7.0.0` following assertions will always pass:
```javascript
reg.test('foo@bar.com').should.be.true;
reg.test('FoO.bAr1@baz.com').should.be.true;
```

Comment (vickvu) [inline test/matchers/email.spec.js]: this project still using shouldjs 6.0

Comment (srushti) [inline test/matchers/email.spec.js]: Yeah, okay. I'm changing it to
```coffee
    reg.test('foo@bar.com').should.eql(true);
```
to play it safe.

File: lib/matchers/email.js (hunk 3 above)
Comment (srushti) [inline lib/matchers/email.js]: Meh. I changed it anyway.

File: lib/matchers/email.js (hunk 2 above)
Comment (srushti) [inline lib/matchers/email.js]: I'm not as sure about this. I don't necessarily go for super-terse with plain js. I think this is clearer. Happy to change it though based on another opinion though.

Comment (mehdivk) [inline lib/matchers/email.js]: What I don't understand is consistency! why we should have three different types of if return syntax in a file, two of them in same function!

You have
```javascript
if (typeof value !== 'string') return TYPE_ERROR;
```

and you have
```javascript
if (value.length > 254)
      return TYPE_ERROR;
```

Also generally speaking this syntax is not accepted in javascript
```javascript
if (value.length > 254)
      return TYPE_ERROR;
```

Take a look at other area of strummer codebase, you will see how if return is used, consistency matters.

Comment (srushti) [inline lib/matchers/email.js]: Yeah, agreed about consistency. I thought you were referring to the `return` below, which is a bit more complicated, and I prefer on a separate line. Fixed the ones above.

**Comment by mehdivk** [comment]
:+1:

**Pull request merged**